### ISSUE TITLE
Empty and Infinite Rectangles

### ISSUE THREAD

**Issue by JorjMcKie** [rk700/PyMuPDF#93]
For version 1.11.0, I have invested some time to clarify MuPDF's notions of "empty" and "infinite" rectangles in our documentation. I hope it is a little clearer know ...

On the other hand, the topic overall is not extremely important, I think.

I investigated Adobe's PDF documentation and did not find corresponding terms - at least no **infinite** rectangles. On the contrary, section 3.8.4 on page 161 of the manual states, that PDF processing applications should be prepared to work with a rectangle, regardless of which of its (diagonally opposing) points are given. This is a clear hint, that PDF knows of no such thing as an infinite rectangle.

Things popped up that might be amazing to users:
* A rectangle can be **empty and infinite** at the same time!
* An empty rectangle is contained in **every** non-empty rectangle, finite or infinite. Well, that makes sense also in mathematics.
* An empty rectangle contains **nothing** - not even itself! And none of its corners ... If this is the right decision in MuPDF could be questioned.
* Rectangle unions where one partner is infinite or empty are always no-operations. This means, `rect.includeRect(fitz.Rect(p,p)) == rect` is always `True` for every point `p`, but in general `rect.includePoint(p) != rect`.

**My questions:**
1. Does anyone know whether other document types, like XPS, could have been the motivator?
2. We easily could do away with infinite rectangles altogether by simple adjustments of our constructors. Would that make sense?

**Enhancements**
Furthermore, MuPDF contains **several bugs** in that area, which I corrected in our PyMuPDF layer. Submitting them has led to no reaction whatsoever ..., so I implemented fixes in PyMuPDF.
Examples:

* `rect.round()` produces a nonsense (empty) `IRect` if `rect` is infinite. PyMuPDF instead now first makes the rectangle finite and calculates the `IRect` on that basis.

* Similarly, `rect.includePoint()` produces a nonsense rectangle if `rect` is infinite. PyMuPDF correction now goes a similar way: first "normalize" `rect` and then include the point.

# Any Feedback?
@ousia ? @deepgully ? @mozbugbox ? @gileadslostson ? ...

**Comment by gileadslostson** [comment]
Sorry, @JorjMcKie! I am in the middle of a project and haven't had a chance to think about it yet.

**Comment by JorjMcKie** [comment]
Never mind - the topic isn't important anyway.

I just stumbled over several points in MuPDF code that are simply unclean or contradictory - and at the same time unnecessary complications.

I still believe that infinite rects **are** an unnecessary notion. Easily avoided during rectangle creation.

While emptiness in contrast **is** a necessary concept, containment should have been implemented differently: an empty rect has no area, allright, but it still **has** points, for which I can state whether they also belong to other rectangles, etc., etc.
